File: backend-ai/app/schemas/match.py
from typing import List, Literal

from pydantic import BaseModel, Field

from app.schemas.job import JobDescriptionInput


class MatchRequest(BaseModel):
    """MVP: paste resume as plain text. File upload is not supported."""

    raw_resume_text: str = Field(..., min_length=1, description="Full resume plain text")
    job: JobDescriptionInput


class MatchResult(BaseModel):
    """Core match payload (no API version). Returned scores are deterministic for fixed input."""

    score: float = Field(..., ge=0, le=100)
    skill_match: float = Field(..., ge=0, le=100)
    experience_match: float = Field(..., ge=0, le=100)
    keyword_ats_match: float = Field(..., ge=0, le=100)
    context_fit: float = Field(..., ge=0, le=100)
    missing_skills: List[str]
    suggestions: List[str]
    weak_areas: List[str] = Field(default_factory=list)


class MatchApiV1(MatchResult):
    version: Literal["v1"] = Field(default="v1", description="API contract version")

File: frontend/src/types/match.ts
import type { JobDescriptionInput } from "./job"

/** MVP: plain-text resume only (no file upload). */
export type MatchRequest = {
  raw_resume_text: string
  job: JobDescriptionInput
}

export type MatchApiV1 = {
  version: "v1"
  score: number
  skill_match: number
  experience_match: number
  keyword_ats_match: number
  context_fit: number
  missing_skills: string[]
  suggestions: string[]
  weak_areas: string[]
}

export type MatchScoreBand = "weak" | "needs_work" | "strong"

export type MatchBandedBucket = {
  score: number
  band: MatchScoreBand
  reasons: string[]
}

export type MatchApiV2 = {
  version: "v2"
  ats_compatibility: MatchBandedBucket
  job_match: MatchBandedBucket
  semantic_similarity: number
  exact_skill_overlap: number
  lexical_match: number
  title_alignment: number
  experience_alignment: number
  missing_hard_skills: string[]
  semantic_matches: string[]
  strengths: string[]
  actions: string[]
  why_this_score: string[]
}

/** Discriminated union for wizard bundle + UI. */
export type MatchPayload = MatchApiV1 | MatchApiV2

export function isMatchV2(m: MatchPayload): m is MatchApiV2 {
  return m.version === "v2"
}
